torch/backends/cudnn/__init__.py

```diff
@@ -1,91 +1,51 @@
 import torch
 import warnings
-from torch.version import cuda
 from contextlib import contextmanager
-from subprocess import Popen, PIPE
 from torch.backends import ContextProp, PropModule, __allow_nonbracketed_mutation
 
+try:
+    from torch._C import _cudnn
+except ImportError:
+    _cudnn = None
+
 # Write:
 #
 #   torch.backends.cudnn.enabled = False
 #
 # to globally disable CuDNN/MIOpen
 
-lib = None
 __cudnn_version = None
-# TODO: dynamic version checks via cudnnGetVersion
-
-def find_cudnn_windows_lib():
-    # Override the default search process
-    # Fixes https://github.com/pytorch/pytorch/issues/20202
-    # The library selection will be done in these directories one by one
-    # 1. [Package Root]\Lib 
-    #    That's where our libraries are in, which should be loaded first.
-    # 2. Default directories
-    #    That is stored in the environment variable `PATH`.
-    test_env = os.environ.copy()
-    old_path = test_env['PATH']
-    th_dll_path = os.path.join(os.path.dirname(
-        os.path.dirname(os.path.dirname(__file__))), 'lib')
-    test_env['PATH'] = ';'.join([th_dll_path, old_path])
-    proc = Popen(['where', 'cudnn64*.dll'], stdout=PIPE,
-                 stderr=PIPE, stdin=PIPE, env=test_env)
-    out, err = proc.communicate()
-    out = out.decode().strip()
-    if len(out) > 0:
-        if out.find('\r\n') != -1:
-            out = out.split('\r\n')[0]
-        cudnn_lib = str(out)
-        return ctypes.cdll.LoadLibrary(cudnn_lib)
-    else:
-        return None
 
-
-def _libcudnn():
-    global lib, __cudnn_version
-    if lib is None:
-        if sys.platform == "win32":
-            lib = find_cudnn_windows_lib()
-        else:
-            lib = ctypes.cdll.LoadLibrary(None)
-        if hasattr(lib, 'cudnnGetErrorString'):
-            lib.cudnnGetErrorString.restype = ctypes.c_char_p
-            __cudnn_version = lib.cudnnGetVersion()
-            compile_version = torch._C._cudnn_version()
-            # cuDNN version is MAJOR*1000 + MINOR*100 + PATCH
-            runtime_major = __cudnn_version // 1000
-            runtime_minor = (__cudnn_version % 1000) // 100
-            compile_major = compile_version // 1000
-            compile_minor = (compile_version % 1000) // 100
+if _cudnn is not None:
+    def _init():
+        global __cudnn_version
+        if __cudnn_version is None:
+            __cudnn_version = _cudnn.getVersionInt()
+            runtime_version = _cudnn.getRuntimeVersion()
+            compile_version = _cudnn.getCompileVersion()
+            runtime_major, runtime_minor, _ = runtime_version
+            compile_major, compile_minor, _ = compile_version
             # Different major versions are always incompatible
             # Starting with cuDNN 7, minor versions are backwards-compatible
             if runtime_major != compile_major:
                 cudnn_compatible = False
-            elif runtime_major < 7:
+            elif runtime_major < 7 or not _cudnn.is_cuda:
```

Comment:
ROCm stuff?

Reply:
Yeah, to be cautious I'm not letting miopen run with a different minor version than the compiled version. But cudnn does apparently allow newer minor versions.

Reply:
also worth a comment!

Reply:
cc @daniellowell in case you have a comment about MIOpen backward-compatibility for minor versions.